---
title: "edgeR Workflow for gene expression PCA and DGE Analysis of *P. acuta*"
author: "Erin Chille"
date: "02/14/2023"
output:
  pdf_document: default
  html_notebook: default
---

```{r setup, include=FALSE}
rm(list = ls()) #clear environment
```

*Note: Open Rproj first, then script. To easily use relative paths, click the down button next to knit and then click "Knit Directory --> Project Directory". This should make loading and saving files much easier.*

## 1. Load packages and input data

Load packages
```{r, message=FALSE, warning=FALSE}
library(edgeR, quietly = TRUE) #edgeR-v3.30.3
library(vegan, quietly = TRUE)
library(Dune, quietly = TRUE)
library(ggplot2, quietly = TRUE) #ggplot2-v3.3.5
library(tidyverse, quietly = TRUE) #tidyverse-v1.3.1
# library(UpSetR, quietly = TRUE)
# library(reshape2, quietly = TRUE)
# library(factoextra, quietly = TRUE)
# library(NbClust, quietly = TRUE)
library(ComplexHeatmap, quietly = TRUE)
library(DESeq2, quietly = TRUE)
library(genefilter, quietly = TRUE)
```

Load the input file containing the treatment information
```{r}
treatmentinfo <- read.csv("Sample_Info/samples_Mcapitata.annotations.txt", header = TRUE, sep = "\t", fileEncoding="UTF-8-BOM") #read in file
```

Load the input file containing the gene count matrix
```{r}
gcount <- as.data.frame(read_delim("Genome_Info/Montipora_capitata_KBHIv3.gentrome.fa.gz.salmon.numreads.matrix", delim = "\t", col_names = TRUE, show_col_types = FALSE), fileEncoding="UTF-8-BOM") #read in file
rownames(gcount) <- gcount$Name #makes "Name" the rowname
gcount <- gcount[,-c(1)] #drops the "Name" column
gcount <- round(gcount) #round 
dim(gcount); head(gcount)[,1:3] #view dataset attributes
gcount <- gcount[ ,treatmentinfo$Sample]
colnames(gcount)==treatmentinfo$Sample
```

Determine library size
```{r}
libSize.df <- data.frame(libSize=colSums(gcount))
```

Make DGE object
```{r}
DGEdat <- DGEList(counts=as.matrix(gcount), samples=treatmentinfo,
                  group=treatmentinfo$Temperature)
dim(DGEdat$counts)
```

## 2. Pre-filtering
```{r}
lib.sizes.cpm <- colSums(gcount)/1000000
lib.sizes.cpm <- data_frame(sample=colnames(gcount), lib.size=lib.sizes.cpm)

ggplot(lib.sizes.cpm, aes(y = lib.size, x=sample)) +
  ylab("Library size (per million)") + xlab("Sample") + # Set x and y axis labels
  ylim(3,10)+
  geom_jitter(colour = "cadetblue", alpha = 0.5) +
  geom_boxplot(colour="cadetblue", alpha=0) + 
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     panel.grid.major = element_blank(), #Set major gridlines
                     panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
                      axis.title.x = element_blank(), # No axis title
                      axis.text.x = element_text(color = "White"), #White-out the x-axis label
                      axis.ticks.x = element_blank(), # No ticks on x-axis
                      axis.title.y = element_text(size = 16), #Set Yaxis label size
                     axis.text=element_text(size=12), #Set Y tick label size
                      plot.background=element_blank()) + #Set the plot background
  ggtitle("(b)") + #add a main title
  scale_color_manual(values=c(AMB="cadetblue", l="indianred3", xl="deeppink4")) + #set treatment colors
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0),
        legend.position = "bottom") #set title attributes
mean(lib.sizes.cpm$lib.size)
median(lib.sizes.cpm$lib.size)
10/min(lib.sizes.cpm$lib.size)

keep <- rowSums(cpm(gcount) > 3.33) >= 2
table(keep)
DGEdat <- DGEdat[keep, , keep.lib.sizes=FALSE]
nrow(DGEdat)
```

##  3. Data normalization  
```{r}
DGEdat <- calcNormFactors(DGEdat)
DGEdat$samples
```

##  4. Plot global gene expression  

Log transform the counts matrix for the next plots
```{r}
DGEdat.cpm <- DGEdat #make a copy the edgeR dataset
DGEdat.cpm$counts <- cpm(DGEdat.cpm$counts, log=TRUE, prior.count=5) #log transform the copy for the next plots
```

Run a principle coordinates analysis
```{r}
d <- data.frame(PC6 = NA, PC7 = NA, Group = DGEdat.cpm$samples$group, DGEdat.cpm$samples, name = colnames(DGEdat.cpm$counts)) #make a dataframe containing all plotting info
pca <- prcomp(t(DGEdat.cpm$counts[,d$Sample])) #calculate eigengenes
pc.data <- summary(out<-prcomp(t(DGEdat.cpm$counts[,d$Sample]))); pc.data
plot(out) #theres a lot of variation in the data
d$Timepoint <- gsub("TP", "", d$Timepoint)
DGEdat_PCcor <- lapply(d[,c(7,9:14)], as.factor)
DGEdat_PCcor <- as.tibble(lapply(DGEdat_PCcor, as.numeric))
DGEdat_PCcor <- cbind(d$lib.size, DGEdat_PCcor)
rownames(DGEdat_PCcor) <- d$Sample
all<-cbind(scores(out), DGEdat_PCcor)
correlations<-round(cor(all, method = "spearman"), digits=2)
plot(correlations)
pc.data
write_csv(as.data.frame(correlations), "Output/1b-Mcap-PC-correlations.csv")

pc_cor<-cbind(scores(out), DGEdat_PCcor)
kruskal.test(PC2 ~ Treatment, data = pc_cor)
kruskal.test(PC2 ~ Temperature, data = pc_cor)
kruskal.test(PC2 ~ CO2, data = pc_cor)

kruskal.test(PC8 ~ Treatment, data = pc_cor)
kruskal.test(PC8 ~ Temperature, data = pc_cor)
kruskal.test(PC8 ~ CO2, data = pc_cor)
# kruskal.test(PC7 ~ treatment, data = pc_cor)
# kruskal.test(PC1 ~ group.1, data = pc_cor)
# # pairwise.wilcox.test(pc_cor$PC1,pc_cor$ploidy,
# #                  p.adjust.method = "BH")
# pairwise.wilcox.test(pc_cor$PC1,pc_cor$group.1,
#                  p.adjust.method = "BH")
```
Visualize first two principle coordinates
```{r}
percentVar <- pca$sdev^2/sum(pca$sdev^2) #save % variation by PC1 and PC2

d <- data.frame(PC2 = pca$x[, 2], PC8 = pca$x[, 8], d) #make a dataframe containing all plotting info

allgenes_PCA <- ggplot(data = d, aes_string(x = "PC2", y = "PC8")) + 
  geom_point(size = 3, color='black', aes(shape=Temperature, fill=Temperature)) +
  scale_shape_manual(guide="legend", values = c('Amb'=21, 'Hot'=24)) +
  scale_fill_manual(guide="legend", values=c('Amb'="#b7b9dc", 'Hot'="#eabaa2")) +
  xlab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) +
  ylab(paste0("PC8: ", round(percentVar[8] * 100), "% variance")) +
  coord_fixed() +
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black", size = 0.6), #Set axes color
        plot.background=element_blank(), #Set the plot background
        axis.title = element_text(size = 14), #Axis title size
        #axis.text = element_blank()
        ); allgenes_PCA #Axis text size and view plot

ggsave("Output/1b-Mcap-edgeR-allsamples-allgenes-PCA.pdf", allgenes_PCA)
```

Plot as a dendrogram
```{r}
tree <- hclust(dist(as.matrix(t(DGEdat.cpm$counts))))
dend <- as.dendrogram(tree)
#pdf("Output/NEW_tree_SNPs_allsamples-allgenes-.pdf", height = 3, width = 20)
plot(dend)
#dev.off()
```

There don't look to be any outliers. 

##  5. Differential gene expression analysis for Montipora capitata with DESeq2

##### Construct the DESeq dataset
Merge the treatment and time_point columns into a new column , group. Set group as a factor.
```{r}
treatmentinfo$Treatment <- factor(treatmentinfo$Treatment, levels = c("ATAC","HTAC"))
treatmentinfo <- filter(treatmentinfo, Treatment%in%c("ATAC", "HTAC"))
treatmentinfo <- filter(treatmentinfo, Timepoint=="1_week")
```

Load the input file containing the gene count matrix
```{r}
gcount <- as.data.frame(read_delim("Genome_Info/Montipora_capitata_KBHIv3.gentrome.fa.gz.salmon.numreads.matrix", delim = "\t", col_names = TRUE, show_col_types = FALSE), fileEncoding="UTF-8-BOM") #read in file
rownames(gcount) <- gcount$Name #makes "Name" the rowname
gcount <- gcount[,-c(1)] #drops the "Name" column
gcount <- round(gcount) #round 
dim(gcount); head(gcount)[,1:3] #view dataset attributes
gcount <- gcount[ ,treatmentinfo$Sample]
colnames(gcount)==treatmentinfo$Sample
```

```{r}
#create filter for the counts data
filt <- filterfun(pOverA(0.5,10))
gfilt <- genefilter(gcount, filt)
#identify genes to keep by count filter
keep <- gcount[gfilt,]
#identify gene lists
keep <- rownames(keep)
#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt <- as.data.frame(gcount[which(rownames(gcount) %in% keep),])
dim(gcount_filt)
```


Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at the interaction of time_point and treatment to test for any differences in gene expression across timepoints attributed to treatment.
```{r}
#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = as.data.frame(gcount_filt),
                              colData = treatmentinfo,
                              design = ~Treatment)
```

#### Differential Gene Expression Analysis

##### Run DE analysis

Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEG <- DESeq(gdds) #run differential expression test by group using the Wald model
```

Explore significant p-values for treatment (HTAC vs ATAC)
```{r, message = FALSE}
DEG.results <- results(DEG, contrast= c("Treatment","HTAC","ATAC"),lfcThreshold=log2(1.5))
head(DEG.results)
sum(DEG.results$padj < 0.05, na.rm=TRUE)
DEGs <- as.data.frame(subset(DEG.results, padj<0.05))
results.ordered <- order(DEGs$padj) #Order p-values by smallest value first
DEGs$gene_id  <- rownames(DEGs)
rownames(DEGs) <- NULL
#write.csv(DEGs, "4-Differential-Gene-Expression-Analysis/Output/DEGs_l_vs_amb.csv")
```

We will now transform them with cpm for plotting
Subset the gene count matrix by the list of DEGs
```{r}
DEGlist <- gdds[DEGs$gene_id,]
```

We know from before that our size factors are all less than 4, so we can use VST!
```{r}
DEGvst <- vst(DEGlist, blind=FALSE, nsub = nrow(counts(DEGlist)), fitType = "local") #apply a variance stabilizing transforamtion to minimize efplncts of small counts and normalize wrt library size
head(assay(DEGvst)) #view transformed gene count data
dim(assay(DEGvst))
```

Make a matrix for computing similarity
```{r}
mat <- assay(DEGvst)#[topVarGenes, ] #make an expression object
mat <- mat - rowMeans(mat) #difference in expression compared to average across all samples
```

Make a heatmap 
```{r}
# mat <- as.matrix(cpm(DEGlist) - rowMeans(cpm(DEGlist)))
hmTreatment <- subset(treatmentinfo, select=c("Temperature"))
hm_ann_col <- HeatmapAnnotation(df=hmTreatment, col = list(Temperature = c("Ambient" = "#1b9e77", "High" = "#d95f02"))) #make dataframe for column naming
dend = cluster_within_group(mat, hmTreatment$Temperature)

DEGheatmap <-  Heatmap(mat, cluster_columns = dend, column_split = 2,
        name = "Gene expression (vst)", 
        show_row_names = T, 
        top_annotation = hm_ann_col, 
        show_column_names = T, row_dend_side = "left" ,
        column_dend_height = unit(0.5, "in"),
        row_title_side = "right", row_title_rot = 0, row_dend_reorder = TRUE, 
        row_gap = unit(2.5, "mm"), border = TRUE,
        column_names_gp =  gpar(fontsize = 10))

pdf("Output/1b-Mcap-edgeR-Treatment-DEG-heatmap.pdf", width = 10.5, height = 7)
draw(DEGheatmap)
dev.off()
```


### Principle components plot of DEGs
```{r}
pca <- prcomp(t(mat)) #calculate eigengenes
percentVar <- pca$sdev^2/sum(pca$sdev^2) #save % variation by PC1 and PC2
d6 <- data.frame(treatmentinfo, PC1 = pca$x[, 1], PC2 = pca$x[, 2])
DEG_PCA <- ggplot(data = d6, aes_string(x = "PC1", y = "PC2")) + 
  geom_point(size = 4, aes(colour=Temperature)) + 
  xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) + 
  ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) + 
  coord_fixed() + 
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black", size = 0.6), #Set axes color
        plot.background=element_blank(), #Set the plot background
        axis.title = element_text(size = 14), #Axis title size
        axis.text = element_blank());DEG_PCA #Axis text size and view plot
ggsave("Output/1b-Mcap-edgeR-allsamples-DEGs-PCA.pdf", DEG_PCA)
```
