---
title: "2b) Pacu Differential Gene Expression analysis treatment within clade"
output:
  pdf_document: default
  html_notebook: default
---

```{r setup, include=FALSE}
rm(list = ls()) #clear environment
```

*Note: Open Rproj first, then script. To easily use relative paths, click the down button next to knit and then click "Knit Directory --> Project Directory". This should make loading and saving files much easier.*

## 1. Load packages and input data

Load packages
```{r, message=FALSE, warning=FALSE}
library(edgeR, quietly = TRUE) #edgeR-v3.30.3
library(vegan, quietly = TRUE)
library(Dune, quietly = TRUE)
library(ggplot2, quietly = TRUE) #ggplot2-v3.3.5
library(tidyverse, quietly = TRUE) #tidyverse-v1.3.1
# library(UpSetR, quietly = TRUE)
# library(reshape2, quietly = TRUE)
# library(factoextra, quietly = TRUE)
# library(NbClust, quietly = TRUE)
library(ComplexHeatmap, quietly = TRUE)
library(DESeq2, quietly = TRUE)
library(genefilter, quietly = TRUE)
```
Differential gene expression analysis for Montipora capitata with DESeq2

##### Construct the DESeq dataset

Load the input file containing the treatment information
```{r}
treatmentinfo <- read.csv("Sample_Info/samples_Pacuta.annotations.txt", header = TRUE, sep = "\t", fileEncoding="UTF-8-BOM") #read in file
treatmentinfo$timepoint <- factor(treatmentinfo$timepoint, levels = c("TP1", "TP3","TP4","TP5","TP6","TP7","TP8","TP9","TP10","TP11"))
top3groups <- c("Group2", "Group3", "Group6")
treatmentinfo <- filter(treatmentinfo, group%in%top3groups)
```

Merge the temp and time_point columns into a new column , group. Set group as a factor.
```{r}
g2.treatmentinfo <- filter(treatmentinfo, group=="Group2")
g2.treatmentinfo$temp <- factor(g2.treatmentinfo$temp, levels = c("Amb","Hot"))
g2.treatmentinfo <- filter(g2.treatmentinfo)
table(g2.treatmentinfo$timepoint, g2.treatmentinfo$temp)


g3.treatmentinfo <- filter(treatmentinfo, group=="Group3")
g3.treatmentinfo$temp <- factor(g3.treatmentinfo$temp, levels = c("Amb","Hot"))
table(g3.treatmentinfo$timepoint, g3.treatmentinfo$temp)

g6.treatmentinfo <- filter(treatmentinfo, group=="Group6")
g6.treatmentinfo$temp <- factor(g6.treatmentinfo$temp, levels = c("Amb","Hot"))
table(g6.treatmentinfo$timepoint, g6.treatmentinfo$temp)
```

We will only conduct DGE analysis for Group 2 samples at TP9 due to limited replicates in other groups and timepoints.

##Group2 DGE Anaylsis

We will do the analysis on TP9, because there are enough replicates at this timepoint for an analysis
```{r}
g2.treatmentinfo <- filter(g2.treatmentinfo, timepoint=="TP9")
```


Load the input file containing the gene count matrix
```{r}
#gene count matrix
gcount <- as.data.frame(read_delim("Genome_Info/Pocillopora_acuta_KBHIv2.gentrome.fa.gz.salmon.numreads.matrix", delim = "\t", col_names = TRUE, show_col_types = FALSE), fileEncoding="UTF-8-BOM") #read in file
rownames(gcount) <- gcount$Name #makes "Name" the rowname
gcount <- gcount[,-c(1)] #drops the "Name" column
gcount <- round(gcount) #round 
dim(gcount); head(gcount)[,1:3] #view dataset attributes
g2.gcount <- gcount[,g2.treatmentinfo$sample]
```

```{r}
#create filter for the counts data
filt <- filterfun(pOverA(0.5,10))
gfilt <- genefilter(g2.gcount, filt)
#identify genes to keep by count filter
keep <- g2.gcount[gfilt,]
#identify gene lists
keep <- rownames(keep)
#gene count data filtered in PoverA, P percent of the samples have counts over A
g2.gcount_filt <- as.data.frame(g2.gcount[which(rownames(g2.gcount) %in% keep),])
dim(g2.gcount_filt)
```


Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at the interaction of time_point and treatment to test for any differences in gene expression across timepoints attributed to treatment.
```{r}
#Set DESeq2 design
gdds <- DESeqDataSetFromMatrix(countData = as.data.frame(g2.gcount_filt),
                              colData = g2.treatmentinfo,
                              design = ~temp)
```

#### Differential Gene Expression Analysis

##### Run DE analysis

Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEG <- DESeq(gdds) #run differential expression test by group using the Wald model
```

Explore significant p-values for temp (Hot vs Amb)
```{r, message = FALSE}
DEG.results <- results(DEG, contrast= c("temp","Hot","Amb"),lfcThreshold=log2(1.5))
head(DEG.results)
sum(DEG.results$padj < 0.05, na.rm=TRUE)
DEGs <- as.data.frame(subset(DEG.results, padj<0.05))
results.ordered <- order(DEGs$padj) #Order p-values by smallest value first
DEGs$gene_id  <- rownames(DEGs)
DEGs$contrast  <- "Group2, Timepoint 9, Hot_v_Amb"
rownames(DEGs) <- NULL
write.csv(DEGs, "Output/1b-Pacu-G2-TP9-Temp-DEGs.csv")
```

We will now transform them with cpm for plotting
Subset the gene count matrix by the list of DEGs
```{r}
topVarGenes <- filter(DEGs, log2FoldChange>2|log2FoldChange<(-2))
DEGlist <- gdds[DEGs$gene_id,]
```

We will now transform them for plotting.
We know from before that our size factors are all less than 4, so we can use VST!
```{r}
DEGvst <- vst(gdds, blind=FALSE, nsub = nrow(counts(gdds)), fitType = "local") #apply a variance stabilizing transforamtion to minimize efplncts of small counts and normalize wrt library size
head(assay(DEGvst)) #view transformed gene count data
dim(assay(DEGvst))
```

Make a matrix for computing similarity
```{r}
mat <- assay(DEGvst)[topVarGenes$gene_id, ] #make an expression object
mat <- mat - rowMeans(mat) #difference in expression compared to average across all samples
```

Make a heatmap 
```{r}
hmTreatment <- subset(g2.treatmentinfo, select=c("temp"))
hm_ann_col <- HeatmapAnnotation(df=hmTreatment, col = list(temp = c("Amb" = "#1b9e77", "Hot" = "#d95f02"))) #make dataframe for column naming
dend = cluster_within_group(mat, hmTreatment$temp)

DEGheatmap <-  Heatmap(mat, cluster_columns = dend, column_split = 2,
        name = "Gene expression (vst)", 
        show_row_names = F, 
        top_annotation = hm_ann_col, 
        show_column_names = T, row_dend_side = "left" ,
        column_dend_height = unit(0.5, "in"),
        row_title_side = "right", row_title_rot = 0, row_dend_reorder = TRUE, 
        row_gap = unit(2.5, "mm"), border = TRUE,
        column_names_gp =  gpar(fontsize = 10)); DEGheatmap

pdf("Output/1b-Pacu-G2-TP9-Temp-DEG-heatmap.pdf", width = 10.5, height = 7)
draw(DEGheatmap)
dev.off()
```

### Principle components plot of DEGs
```{r}
pca <- prcomp(t(mat)) #calculate eigengenes
percentVar <- pca$sdev^2/sum(pca$sdev^2) #save % variation by PC1 and PC2
d6 <- data.frame(g2.treatmentinfo, PC1 = pca$x[, 1], PC2 = pca$x[, 2])
DEG_PCA <- ggplot(data = d6, aes_string(x = "PC1", y = "PC2")) + 
  geom_point(size = 4, aes(colour=temp)) + 
  xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) + 
  ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) + 
  coord_fixed() + 
  theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black", size = 0.6), #Set axes color
        plot.background=element_blank(), #Set the plot background
        axis.title = element_text(size = 14), #Axis title size
        axis.text = element_blank());DEG_PCA #Axis text size and view plot
# ggsave("Output/1b-Pacu-G2-TP9-Temp-DEGs-PCA.pdf", DEG_PCA)
``