---
title: "Gene Ontology Categorization of DEGs between Pacuta Groups"
author: "Erin Chille"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list = ls()) #clear environment
```

---

## RNAseq Differential Expression Analysis 

### Set up workspace

Load libraries
```{r, message=FALSE, warning=FALSE}
library("tidyverse")
library("RColorBrewer")
library("ggplot2")
library("goseq")
library("gridExtra")
library("patchwork")
library("BiocParallel")
library("GSEABase")
library("data.table")
library("cowplot")
```

## GO Enrichment Analysis

Load transcript annotations and transcript length information
```{r}
cluster.df <- read_csv( "Output/2a-Pacu-group-DEGs.csv", show_col_types = FALSE)
cluster.df <- subset(cluster.df, select=c("gene_id", "contrast"))

GO.df <- read_tsv("Genome_Info/Pocillopora_acuta_KBHIv2.genes.EggNog_results.txt", show_col_types = FALSE)
GO.df <- tibble("gene_id"=GO.df$`#query`, "GO_terms"=GO.df$GOs); head(GO.df)

length.df <- read_tsv("Genome_Info/Pocillopora_acuta_KBHIv2.gentrome.fa.gz.salmon.elen.matrix.txt", show_col_types = TRUE)
length.df <- tibble("gene_id"=length.df$Name, "mean_length"=round(rowMeans(length.df[,-1]))); head(length.df)
```

Separate out different gene groups
```{r}
#Triploids v Diploids (only shared genes)
#Up
G2vG6 <- cluster.df %>% filter(contrast == "Up.Group2_v_Group6")
G3vG6 <- cluster.df %>% filter(contrast == "Up.Group3_v_Group6")
TvD.shared.up <- intersect(G2vG6$gene_id, G3vG6$gene_id)
Up.TvD <- filter(cluster.df, gene_id %in% TvD.shared.up)

#Down
G2vG6 <- cluster.df %>% filter(contrast == "Down.Group2_v_Group6")
G3vG6 <- cluster.df %>% filter(contrast == "Down.Group3_v_Group6")
TvD.shared.down <- intersect(G2vG6$gene_id, G3vG6$gene_id)
Down.TvD <- filter(cluster.df, gene_id %in% TvD.shared.down)

intersect(Up.TvD$gene_id, Down.TvD$gene_id) #Should be no overlap

#Between triploid groups
#Up
Up.G3vG2 <- filter(cluster.df, contrast == "Up.Group3_v_Group2")

#Down
Down.G3vG2 <- filter(cluster.df, contrast == "Down.Group3_v_Group2")
```

Form a reference dataframe containing all relevant gene information, including gene_id, associated GO_terms, and gene length.
```{r}
# GO.ref <- right_join(cluster.df, GO.df, by="gene_id")
GO.ref <- right_join(GO.df, length.df, by="gene_id")
GO.ref$GO_terms <- gsub("-","Unknown",GO.ref$GO_terms)
# GO.ref$GO_terms[GO.ref$GO_terms == "-"] <- NA
# GO.ref <- na.omit(GO.ref)
str(GO.ref) 
```

Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
```{r}
# GO.ref <- str(unique(GO.ref))
IDvector <- GO.ref$gene_id
lengthVector <- GO.ref$mean_length

#Cluster1 -- Up in triploids

C1genes <- as.vector(unique(Up.TvD$gene_id))
length(C1genes)
C1genes.matrix=as.integer(GO.ref$gene_id%in%C1genes)
names(C1genes.matrix)=GO.ref$gene_id
length(C1genes.matrix)

#Cluster2 -- Down in triploids

C2genes <- as.vector(unique(Down.TvD$gene_id))
length(C2genes)
C2genes.matrix=as.integer(GO.ref$gene_id%in%C2genes)
names(C2genes.matrix)=GO.ref$gene_id
length(C2genes.matrix)

#Cluster3 -- Up in Group3 (v Group2)

C3genes <- as.vector(unique(Up.G3vG2$gene_id))
length(C3genes)
C3genes.matrix=as.integer(GO.ref$gene_id%in%C3genes)
names(C3genes.matrix)=GO.ref$gene_id
length(C3genes.matrix)

#Cluster4 -- Down in Group3 (v Group2)

C4genes <- as.vector(unique(Down.G3vG2$gene_id))
length(C4genes)
C4genes.matrix=as.integer(GO.ref$gene_id%in%C4genes)
names(C4genes.matrix)=GO.ref$gene_id
length(C4genes.matrix)
```

Run a probability weighting function using the vectors and matrices we made above for each group.
```{r, message=FALSE}
pwf.1<-nullp(DEgenes=C1genes.matrix, id=IDvector, bias.data=lengthVector)
pwf.2<-nullp(DEgenes=C2genes.matrix, id=IDvector, bias.data=lengthVector)
pwf.3<-nullp(DEgenes=C3genes.matrix, id=IDvector, bias.data=lengthVector)
pwf.4<-nullp(DEgenes=C4genes.matrix, id=IDvector, bias.data=lengthVector)
```

Stopping to check that the matrix worked as it should have. Checking the number of DEGs
```{r}
nrow(filter(pwf.1, DEgenes==1)) == length(C1genes)
nrow(filter(pwf.2, DEgenes==1)) == length(C2genes)
nrow(filter(pwf.3, DEgenes==1)) == length(C3genes)
nrow(filter(pwf.4, DEgenes==1)) == length(C4genes)
```

Prepare GO term dataframe
```{r}
GO.annot <- subset(GO.ref, select=c(gene_id, GO_terms))

splitted <- strsplit(as.character(GO.annot$GO_terms), ",") #split into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GO.annot$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO.terms) <- c("gene_id", "GO.ID")

GO.terms$GO.ID<- as.character(GO.terms$GO.ID)
# GO.terms$GO.ID <- replace_na(GO.terms$GO.ID, "unknown")
GO.terms$GO.ID <- as.factor(GO.terms$GO.ID)
GO.terms$gene_id <- as.factor(GO.terms$gene_id)
GO.terms$GO.ID <- gsub(" ", "", GO.terms$GO.ID)
GO.terms <- unique(GO.terms)

dim(GO.terms)
head(GO.terms)
```

Find enriched GO terms, “selection-unbiased testing for category enrichment among significantly expressed genes for RNA-seq data”
```{r, message=FALSE}
GOwall.1 <- goseq(pwf.1, IDvector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GOwall.2 <- goseq(pwf.2, IDvector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GOwall.3 <- goseq(pwf.3, IDvector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GOwall.4 <- goseq(pwf.4, IDvector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
```
Find only enriched GO terms that are statistically significant at cutoff
```{r}
up.GO1<-GOwall.1$category[GOwall.1$over_represented_pvalue<.05]
up.GO1<-data.frame(up.GO1)
colnames(up.GO1) <- c("category")
up.GO1 <- merge(up.GO1, GOwall.1, by="category")
up.GO1 <- up.GO1[order(up.GO1$ontology, up.GO1$over_represented_pvalue, -up.GO1$numDEInCat),]
up.GO1$term <- as.factor(up.GO1$term)
nrow(filter(up.GO1, ontology=="BP")) #number sig BP terms
nrow(filter(up.GO1, ontology=="MF")) #number sig MF terms
nrow(up.GO1)

up.GO2<-GOwall.2$category[GOwall.2$over_represented_pvalue<.05]
up.GO2<-data.frame(up.GO2)
colnames(up.GO2) <- c("category")
up.GO2 <- merge(up.GO2, GOwall.2, by="category")
up.GO2 <- up.GO2[order(up.GO2$ontology, up.GO2$over_represented_pvalue, -up.GO2$numDEInCat),]
up.GO2$term <- as.factor(up.GO2$term)
nrow(filter(up.GO2, ontology=="BP")) #number sig BP terms
nrow(filter(up.GO2, ontology=="MF")) #number sig MF terms
nrow(up.GO2)

up.GO3<-GOwall.3$category[GOwall.3$over_represented_pvalue<.05]
up.GO3<-data.frame(up.GO3)
colnames(up.GO3) <- c("category")
up.GO3 <- merge(up.GO3, GOwall.3, by="category")
up.GO3 <- up.GO3[order(up.GO3$ontology, up.GO3$over_represented_pvalue, -up.GO3$numDEInCat),]
up.GO3$term <- as.factor(up.GO3$term)
nrow(filter(up.GO3, ontology=="BP")) #number sig BP terms
nrow(filter(up.GO3, ontology=="MF")) #number sig MF terms
nrow(up.GO3)

up.GO4<-GOwall.4$category[GOwall.4$over_represented_pvalue<.05]
up.GO4<-data.frame(up.GO4)
colnames(up.GO4) <- c("category")
up.GO4 <- merge(up.GO4, GOwall.4, by="category")
up.GO4 <- up.GO4[order(up.GO4$ontology, up.GO4$over_represented_pvalue, -up.GO4$numDEInCat),]
up.GO4$term <- as.factor(up.GO4$term)
nrow(filter(up.GO4, ontology=="BP")) #number sig BP terms
nrow(filter(up.GO4, ontology=="MF")) #number sig MF terms
nrow(up.GO4)
```

## Find associated GOslim terms to streamline interpretation of our very cool results

Join our dataframes containing significantly up and down-regulated terms.
```{r}
up.GO1$Cluster <- "Up in Triploids"
up.GO2$Cluster <- "Down in Triploids"
up.GO3$Cluster <- "Up in Group3 v Group2"
up.GO4$Cluster <- "Down in Group3 v Group2"
all_GO <- bind_rows(up.GO1, up.GO2, up.GO3, up.GO4)  #bind rows
str(all_GO)
```
Run GOslim to get our umbrella terms.
```{r}
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database

## BP
BP_GO <- all_GO %>%
  filter(ontology=="BP")
BPGO_collection <- GOCollection(BP_GO$category) #Make library of query terms
slims_bp <- data.frame(goSlim(BPGO_collection, slim, "BP")) #Find common parent terms to slim down our list
slims_bp$category <- row.names(slims_bp) #save rownames as category

## MF
MF_GO <- all_GO %>%
  filter(ontology=="MF")
MFGO_collection <- GOCollection(MF_GO$category) #Make library of query terms
slims_mf <- data.frame(goSlim(MFGO_collection, slim, "MF")) #Find common parent terms to slim down our list
slims_mf$category <- row.names(slims_mf) #save rownames as category
```

Get mapped terms, using functions from Sam White’s Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r}
#Write function mappedIds to get the query terms that mapped to the slim categories
mappedIds <-
  function(df, collection, OFFSPRING) #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  {
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
#Run function for MF and BP terms
BPslim <- mappedIds(slims_bp, BPGO_collection, GOBPOFFSPRING)
MFslim <- mappedIds(slims_mf, MFGO_collection, GOMFOFFSPRING)
```

Remove duplicate matches, keeping the broader umbrella term
```{r}
#BP
BPslim <- filter(BPslim, Count>0 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(BPslim$go_terms), ";") #split into multiple GO ids
BPslimX <- data.frame(Term = rep.int(BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
BPslimX <- merge(BPslimX, BPslim[,c(1,3:4)], by="Term") #Add back counts, term, and category info
BPslimX <- unique(setDT(BPslimX)[order(go_term, -Count)], by = "go_term") #remove duplicate offspring terms, keeping only those that appear in the larger umbrella term (larger Count number)
BPslim <- data.frame(slim_term=BPslimX$Term, slim_cat=BPslimX$category, category=BPslimX$go_term) #rename columns
head(BPslim)

#MF
MFslim <- filter(MFslim, Count>0 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(MFslim$go_terms), ";") #split into multiple GO ids
MFslimX <- data.frame(Term = rep.int(MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all
MFslimX <- merge(MFslimX, MFslim[,c(1,3:4)], by="Term")  #Add back counts, term, and category info
MFslimX <- unique(setDT(MFslimX)[order(go_term, -Count)], by = "go_term")  #remove duplicate offspring terms, keeping only
MFslim <- data.frame(slim_term=MFslimX$Term, slim_cat=MFslimX$category, category=MFslimX$go_term) #rename columns
head(MFslim)
```

Save slim info with GO enrichment info for heatmap dataframes.
```{r}
GO.BP <- right_join(BPslim, filter(all_GO, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
GO.MF <- right_join(MFslim, filter(all_GO, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term
```

### Make a heatmap to visualize our functional enrichment results
```{r}
GO.BP$Cluster <- factor(GO.BP$Cluster, levels=c("Up in Triploids", "Down in Triploids", "Up in Group3 v Group2", "Down in Group3 v Group2"))
BPplot <- na.omit(GO.BP) %>% slice_max(n=5, by = slim_cat, order_by = over_represented_pvalue) %>% 
  mutate(term = fct_reorder(term, -over_represented_pvalue)) %>% ggplot(aes(x = Cluster, y = term)) + 
  geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
  facet_grid(slim_term ~ ontology, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
  strip.text.y = element_text(angle=0, size = 6, face = "bold"),
  strip.text.x = element_text(size = 6, face = "bold"),
  axis.title.x = element_blank(),
  axis.title.y = element_text(size=7),
  axis.text = element_text(size = 6), legend.position = "None",
  axis.text.x = element_text(angle = 90),
  plot.margin = unit(c(0,1,0,0.25), "cm"));BPplot
ggsave("Output/3-Pacu-group-GOslimBPplot.pdf", BPplot, width = 10, height = 20, units = c("in"))
```

## Make supplemental table summarizing GO enrichment

Must do separately for C1 (u)p and C2 (down) so we don't get duplicate gene entries.
```{r}
DF_GOhm <- bind_rows(GO.BP, GO.MF)
head(DF_GOhm)



#Separate up and down
DF_GOhm_U1 <- filter(DF_GOhm, Cluster=="Up in Triploids")
DF_GOhm_D1 <- filter(DF_GOhm, Cluster=="Down in Triploids")
DF_GOhm_U2 <- filter(DF_GOhm, Cluster=="Up in Group3 v Group2")
DF_GOhm_D2 <- filter(DF_GOhm, Cluster=="Down in Group3 v Group2")
```

Make dataframe of GO results for clustering and heatmap. 
```{r}
#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOgenes <- data.frame(gene_id=GO.terms$gene_id, category=GO.terms$GO.ID) #First have to make the "by" column the same for both
#GOgenes$gene_id <- as.character(GOgenes$gene_id)
```

Using R script from Mass Lab. Needs columns "experiment" where we will put cluster information, "term", "percentDEInCat", and "gene" with all the DE genes associated with that GO term.
```{r}
#DO TRIPLOID V DIPLOID
GOgenes_U1 <- filter(GOgenes, gene_id%in%C1genes)
GOdf_U1 <- left_join(DF_GOhm_U1, GOgenes_U1, by="category" ) #join the DFs
GOdf_U1$gene_id <- as.character(GOdf_U1$gene_id) #make gene ID a character so we can collapse our many near-identical columns
GOdf_U1 <- GOdf_U1 %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  filter(gene_id%in%C1genes) %>% 
  group_by(ontology, slim_term, slim_cat, category, term, over_represented_pvalue, numDEInCat, numInCat, Cluster) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()

GOgenes_D1 <- filter(GOgenes, gene_id%in%C2genes)
GOdf_D1 <- left_join(DF_GOhm_D1, GOgenes_D1, by="category" ) #join the DFs
GOdf_D1$gene_id <- as.character(GOdf_D1$gene_id) #make gene ID a character so we can collapse our many near-identical columns
GOdf_D1 <- GOdf_D1 %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  filter(gene_id%in%C2genes) %>% 
  group_by(ontology, slim_term, slim_cat, category, term, over_represented_pvalue, numDEInCat, numInCat, Cluster) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()

#DO GROUP3 V GROUP2
GOgenes_U2 <- filter(GOgenes, gene_id%in%C3genes)
GOdf_U2 <- left_join(DF_GOhm_U2, GOgenes_U2, by="category" ) #join the DFs
GOdf_U2$gene_id <- as.character(GOdf_U2$gene_id) #make gene ID a character so we can collapse our many near-identical columns
GOdf_U2 <- GOdf_U2 %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  filter(gene_id%in%C3genes) %>% 
  group_by(ontology, slim_term, slim_cat, category, term, over_represented_pvalue, numDEInCat, numInCat, Cluster) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()

GOgenes_D2 <- filter(GOgenes, gene_id%in%C4genes)
GOdf_D2 <- left_join(DF_GOhm_D2, GOgenes_D2, by="category" ) #join the DFs
GOdf_D2$gene_id <- as.character(GOdf_D2$gene_id) #make gene ID a character so we can collapse our many near-identical columns
GOdf_D2 <- GOdf_D2 %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  filter(gene_id%in%C4genes) %>% 
  group_by(ontology, slim_term, slim_cat, category, term, over_represented_pvalue, numDEInCat, numInCat, Cluster) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()
```

Bind rows for up and down and save!
```{r}
GOdf <- bind_rows(GOdf_U1, GOdf_D1, GOdf_U2, GOdf_D2)
head(GOdf)
str(GOdf)
write.csv(GOdf, file = "Output/3-Pacu-group-DEG-GOenrichment-results.csv", row.names = FALSE)
```


